### PR TITLE
Add unapproving to data model

### LINKED ISSUE #108: Update Data Model to Support Unapproving
## Overview

A Validator or Administrator can unapprove a submission, moving it back to Submitted. This should be reflected in the Activity Log.

Figure out the best way to implement this in our data model.

### PR DESCRIPTION
**Pull request by mstone121**
## Overview

This PR mainly adds unapproving to the data model. In order to do this, approvals are converted from a One to One, to Many to One (Foreign key) relationship.

Additionally this PR updates the activity log serializer to make its operation more transparent.

Closes #108 

### Notes

The Activity Log updates happened here for a couple of reasons:
- Multiple approvals for a single submission means the approval entries can't be serialized from the submission without returning multiple events. Instead, it seemed to make more sense to pass just the object which contains the event data to the serializer and go from there.
- Since reviews might happen after approvals, we can't always add them after and expect the activity log to be sorted. Luckily, all events have a common `time` property we can sort on. This allowed me to re-order the activity log generation to read in order of the most common submission process. 

## Testing Instructions

- `./scripts/resetdb` or `./scripts/manage migrate`
- UI elements to be added in #151, so testing here is limited to making sure boundary list and details pages load correctly with special attention given to the activity log.

## Checklist

- [ ] `fixup!` commits have been squashed
- [ ] `CHANGELOG.md` updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [ ] `README.md` updated if necessary to reflect the changes
- [ ] CI passes after rebase

### CODE REVIEW

File: src/django/api/models/submission.py [hunk 1]
```diff
@@ -75,9 +75,26 @@ class Approval(models.Model):
...
+        User,
+        on_delete=models.PROTECT,
+        null=True,
+        blank=True,
+        limit_choices_to=limit_by_validator_or_admin,
+        related_name='unapprover',
     )
 
+    def clean(self):
+        if self.unapproved_at is not None and self.approved_by is None:
```
Comment (rajadain) [inline src/django/api/models/submission.py:105]:
```suggestion
        if self.unapproved_at is not None and self.unapproved_by is None:
```

**Review by rajadain — CHANGES_REQUESTED** [review]
Would recommend adding an example of unapproval to the test dataset:

```diff
diff --git a/src/django/api/tests.py b/src/django/api/tests.py
index 5d65a13..8798da3 100644
--- a/src/django/api/tests.py
+++ b/src/django/api/tests.py
@@ -247,7 +247,7 @@ class BoundarySyncAPITestCase(TestCase):
         )
 
         # approved
-        Approval.objects.create(
+        approval_6 = Approval.objects.create(
             submission=resubmission_6,
             approved_at=datetime.now(tz=timezone.utc),
             approved_by=cls.validator,
@@ -259,6 +259,12 @@ class BoundarySyncAPITestCase(TestCase):
             approved_by=cls.validator,
         )
 
+        # Unapprove
+
+        approval_6.unapproved_at = datetime.now(tz=timezone.utc)
+        approval_6.unapproved_by = cls.validator
+        approval_6.save()
+
         cls.reference_image = ReferenceImage.objects.create(
             filename="test_file.jpg",
             is_visible=True,
```

Doing this reveals that the front-end needs to be updated at least in one place:

```diff
diff --git a/src/app/src/components/Submissions/ActivityLog.js b/src/app/src/components/Submissions/ActivityLog.js
index d5243e0..1f86352 100644
--- a/src/app/src/components/Submissions/ActivityLog.js
+++ b/src/app/src/components/Submissions/ActivityLog.js
@@ -58,6 +58,8 @@ function getActionString(action) {
             return ' requested revisions.';
         case BOUNDARY_STATUS.APPROVED:
             return ' approved this map.';
+        case 'Unapproved':
+            return ' unapproved this map.';
 
         default:
             throw new Error(`Invalid action type: ${action}`);
```

File: src/django/api/models/submission.py (hunk 1 above)
Comment (mstone121) [reply src/django/api/models/submission.py]: [e205785](https://github.com/azavea/iow-boundary-tool/pull/178/commits/e205785d6f3678249e5e358f9bb4211ec3b8f474)

**Comment by mstone121** [comment]
Ready for another look.

Activity log update: [220ca2e](https://github.com/azavea/iow-boundary-tool/pull/178/commits/220ca2e12dd5dc2f0db927419f51a7ffd0cd1897)
Test data: https://github.com/azavea/iow-boundary-tool/pull/178/commits/d30b3541dc4b84ded7625712ee1ea0df4069a8f7

**Comment by rajadain** [comment]
Heads up, this will require moving the test updates to the new location, and also for regenerating the migrations to be in the right order, to be on top of #175.

**Comment by mstone121** [comment]
Post rebase checks:
- [x] Test updates moved to base.py (`git` did this automatically)
- [x] `./scripts/resetdb` finishes
- [x] App loads and unapproved test data shows up correctly

I'll merge once CI finishes.

**Pull request merged**